turbopack/crates/turbopack-core/src/module_graph/mod.rs

```diff
@@ -1274,20 +863,87 @@ impl ModuleGraphRef {
+        let graph = &self.graphs[entry.graph_idx()];
+        graph
+            .graph
+            .node_weight(entry.node_idx)
+            .context("Expected graph node")
+    }
+
     /// Returns a map of all modules in the graphs to their identifiers.
     /// This is primarily useful for debugging.
     pub async fn get_ids(&self) -> Result<FxHashMap<ResolvedVc<Box<dyn Module>>, ReadRef<RcStr>>> {
         Ok(self
             .graphs
             .iter()
             .flat_map(|g| g.iter_nodes())
-            .map(async |n| Ok((n.module, n.module.ident().to_string().await?)))
+            .map(async |n| Ok((n, n.ident().to_string().await?)))
             .try_join()
             .await?
             .into_iter()
             .collect::<FxHashMap<_, _>>())
     }
 
+    /// Traverses all reachable nodes exactly once and calls the visitor.
+    ///
+    /// * `entry` - The entry module to start the traversal from
+    /// * `visitor` - Called before visiting the children of a node.
+    ///    - Receives &SingleModuleGraphNode
+    ///    - Can return [GraphTraversalAction]s to control the traversal
+    pub fn traverse_nodes_from_entries_dfs<S>(
+        &self,
+        entries: impl IntoIterator<Item = ResolvedVc<Box<dyn Module>>>,
+        state: &mut S,
+        visit_preorder: impl Fn(ResolvedVc<Box<dyn Module>>, &mut S) -> Result<GraphTraversalAction>,
+        mut visit_postorder: impl FnMut(ResolvedVc<Box<dyn Module>>, &mut S) -> Result<()>,
+    ) -> Result<()> {
+        let graphs = &self.graphs;
+        let skip_visited_module_children = self.skip_visited_module_children;
+
+        let entries = entries.into_iter().collect::<Vec<_>>();
+
+        enum Pass {
+            Visit,
+            ExpandAndVisit,
+        }
+        let mut stack: Vec<(Pass, GraphNodeIndex)> = Vec::with_capacity(entries.len());
+        for entry in entries.into_iter().rev() {
+            stack.push((Pass::ExpandAndVisit, self.get_entry(entry)?));
+        }
+        let mut expanded = FxHashSet::default();
+        while let Some((pass, current)) = stack.pop() {
+            let current_node = self.get_node(current)?;
+            match pass {
+                Pass::Visit => {
+                    visit_postorder(current_node.module(), state)?;
+                }
+                Pass::ExpandAndVisit => {
+                    if !expanded.insert(current) {
+                        continue;
+                    }
+                    let action = visit_preorder(current_node.module(), state)?;
+                    if action == GraphTraversalAction::Exclude {
+                        continue;
+                    }
+                    stack.push((Pass::Visit, current));
+                    if action == GraphTraversalAction::Continue
+                        && !(skip_visited_module_children
+                            && matches!(current_node, SingleModuleGraphNode::VisitedModule { .. }))
```

Comment:
nit:expose as a private method on `ModuleGraphRef`  this pattern is repeated in a few places